drivers/sensor/lsm6dsr/lsm6dsr.h

```diff
@@ -0,0 +1,98 @@
+/* ST Microelectronics LSM6DSO 6-axis IMU sensor driver
+ *
+ * Copyright (c) 2019 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Datasheet:
+ * https://www.st.com/resource/en/datasheet/lsm6dsr.pdf
+ */
+
+#ifndef ZEPHYR_DRIVERS_SENSOR_LSM6DSR_LSM6DSR_H_
+#define ZEPHYR_DRIVERS_SENSOR_LSM6DSR_LSM6DSR_H_
+
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/types.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <stmemsc.h>
+#include "lsm6dsr_reg.h"
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/i2c.h>
+
+#define LSM6DSR_EN_BIT					0x01
+#define LSM6DSR_DIS_BIT					0x00
+
+/* Accel sensor sensitivity grain is 61 ug/LSB */
+#define SENSI_GRAIN_XL				(61LL)
+
+/* Gyro sensor sensitivity grain is 4375 udps/LSB */
+#define SENSI_GRAIN_G				(4375LL)
+#define SENSOR_PI_DOUBLE			(SENSOR_PI / 1000000.0)
+#define SENSOR_DEG2RAD_DOUBLE		(SENSOR_PI_DOUBLE / 180)
+#define SENSOR_G_DOUBLE				(SENSOR_G / 1000000.0)
+
+struct lsm6dsr_config {
+	stmdev_ctx_t ctx;
+	union {
+		const struct i2c_dt_spec i2c;
+		const struct spi_dt_spec spi;
+	} stmemsc_cfg;
+	uint8_t accel_pm;
+	uint8_t accel_odr;
+	uint8_t accel_range;
+	uint8_t gyro_pm;
+	uint8_t gyro_odr;
+	uint8_t gyro_range;
+	bool drdy_pulsed;
+#ifdef CONFIG_LSM6DSR_TRIGGER
+	const struct gpio_dt_spec gpio_drdy;
+	uint8_t int_pin;
+	bool trig_enabled;
+#endif /* CONFIG_LSM6DSR_TRIGGER */
+};
+
+struct lsm6dsr_data {
+	const struct device *dev;
+	int16_t acc[3];
+	uint32_t acc_gain;
+	uint8_t accel_pm;
+	int16_t gyro[3];
+	uint32_t gyro_gain;
+	int16_t temp_sample;
+
+	float accel_freq;
+	uint8_t accel_fs;
+	float gyro_freq;
+	uint8_t gyro_fs;
+
+#ifdef CONFIG_LSM6DSR_TRIGGER
```

Comment:
Could be a good option to put this part into a separate structure - for example `struct lsm6dsr_trigger_data`.

Reply:
To be honest, I'm not really sure how this supposed to be done, does the following patch resolves this issue?
```patch
-#define LSM6DSR_DEFINE(inst)				\
-	static struct lsm6dsr_data lsm6dsr_data_##inst;			\
-	static const struct lsm6dsr_config lsm6dsr_config_##inst =	\
-		COND_CODE_1(DT_INST_ON_BUS(inst, spi),			\
-			(LSM6DSR_CONFIG_SPI(inst)),			\
-			(LSM6DSR_CONFIG_I2C(inst)));			\
+#define LSM6DSR_DEFINE(inst)								\
+	IF_ENABLED(CONFIG_LSM6DSR_TRIGGER,						\
+		(static struct lsm6dsr_trigger_data lsm6dsr_trigger_data_##inst;)) 	\
+	static struct lsm6dsr_data lsm6dsr_data_##inst = {				\
+		IF_ENABLED(CONFIG_LSM6DSR_TRIGGER,					\
+			(.trig_data = &lsm6dsr_trigger_data_##inst,))			\
+	};										\
+	static const struct lsm6dsr_config lsm6dsr_config_##inst =			\
+		COND_CODE_1(DT_INST_ON_BUS(inst, spi),					\
+			(LSM6DSR_CONFIG_SPI(inst)),					\
+			(LSM6DSR_CONFIG_I2C(inst)));					\
```

Reply:
This could be the structures split:
```c
struct lsm6dsr_trigger_data {
	struct gpio_callback gpio_cb;
	/* ... */

#if defined(CONFIG_LSM6DSR_TRIGGER_OWN_THREAD)
	K_KERNEL_STACK_MEMBER(thread_stack, CONFIG_LSM6DSR_THREAD_STACK_SIZE);
	struct k_thread thread;
	struct k_sem gpio_sem;
#elif defined(CONFIG_LSM6DSR_TRIGGER_GLOBAL_THREAD)
	struct k_work work;
#endif
};
```
```c
struct lsm6dsr_data {
	const struct device *dev;
	/* ... */

#if defined(CONFIG_LSM6DSR_TRIGGER)
	struct lsm6dsr_trigger_data *trig_data;
#endif
};
```

Reply:
I think that should work? Sorry this has slipped past my radar for some time. I'll resolve my comment.